syntax = "proto3";

enum OrderbookLevelUpdateType {
    INVALID = 0;
    ADD = 1;
    REPLACE = 2;
    REMOVE = 3;
}

message OrderbookLevel {
    int32 price = 1;
    bool is_buy = 2;
    uint32 quantity = 3;
}

message OrderbookLevelUpdate {
    OrderbookLevelUpdateType update_type = 1;
    OrderbookLevel level = 2;
}

message OrderbookSnapshotUpdate {
    repeated OrderbookLevel bids = 1;
    repeated OrderbookLevel asks = 2;
}

message OrderbookUpdate {
    int32 instrument_id = 1;
    oneof update {
        OrderbookSnapshotUpdate snapshot = 2;
        OrderbookLevelUpdate incremental = 3;
    }
}

// Missing message and service definitions

// Subscription related messages
message InstrumentIds {
    repeated int32 ids = 1;
}

message Subscription {
    oneof action {
        InstrumentIds subscribe = 1;
        InstrumentIds unsubscribe = 2;
    }
}

// Service definition
service MarketDataService {
    // Bidirectional streaming RPC for orderbook updates
    rpc StreamOrderbookUpdates(stream Subscription) returns (stream OrderbookUpdate) {}
}